helix-vcs/src/lib.rs

```diff
@@ -20,44 +21,26 @@ pub trait DiffProvider {
     /// The data is returned as raw byte without any decoding or encoding performed
     /// to ensure all file encodings are handled correctly.
     fn get_diff_base(&self, file: &Path) -> Result<Vec<u8>>;
-    fn get_current_head_name(&self, file: &Path) -> Result<Arc<ArcSwap<Box<str>>>>;
 }
 
-#[doc(hidden)]
-pub struct Dummy;
-impl DiffProvider for Dummy {
-    fn get_diff_base(&self, _file: &Path) -> Result<Vec<u8>> {
-        bail!("helix was compiled without git support")
-    }
-
-    fn get_current_head_name(&self, _file: &Path) -> Result<Arc<ArcSwap<Box<str>>>> {
-        bail!("helix was compiled without git support")
-    }
+pub trait VcsProvider {
+    fn root_for_file(&self, file: &Path) -> Result<Box<dyn VcsRoot>>;
 }
 
-pub struct DiffProviderRegistry {
-    providers: Vec<Box<dyn DiffProvider>>,
+pub trait VcsRoot: DiffProvider {
+    fn get_current_head_name(&self) -> Result<Arc<ArcSwap<Box<str>>>>;
 }
 
-impl DiffProviderRegistry {
-    pub fn get_diff_base(&self, file: &Path) -> Option<Vec<u8>> {
-        self.providers
-            .iter()
-            .find_map(|provider| match provider.get_diff_base(file) {
-                Ok(res) => Some(res),
-                Err(err) => {
-                    log::info!("{err:#?}");
-                    log::info!("failed to open diff base for {}", file.display());
-                    None
-                }
-            })
-    }
+pub struct VcsProviderRegistry {
+    providers: Vec<Box<dyn VcsProvider>>,
+}
 
-    pub fn get_current_head_name(&self, file: &Path) -> Option<Arc<ArcSwap<Box<str>>>> {
+impl VcsProviderRegistry {
+    pub fn root_for_file(&self, file: &Path) -> Option<Box<dyn VcsRoot>> {
```

Comment:
should the `DiffProviderRegistry` just be removed entirely in that case? the reason i structured it this way is that i had thought that it would be useful to be able to try multiple vcs providers to figure out which one applied to the current directory, but then be able to diff against the one that it found in various different ways, but if you want to remove the fallback behavior entirely, i can also do that

Reply:
yeah exactly I wnat to remove the fallback behaviour entirely